File: C++/Termostato.cpp
#include <iostream>
using namespace std; // estava faltando isso

class Termostato
{
private:
    int temperatura;
    bool ligado;
    const int MAX = 40;
    const int MIN = 0;

    // CORREÇÃO: a lógica estava invertida
    // retorna true se for VÁLIDO, false se estiver fora dos limites
    bool validar_Temperatura(int valor){
        if (valor < MIN || valor > MAX)
        {
            cout << "Aviso: Temperatura fora dos limites permitidos (0 a 40)!" << endl;
            return false; // inválido
        }
        return true; // válido
    }

public:

    // CORREÇÃO: construtor ignorava os parâmetros e usava os nomes errados
    // this-> é necessário para diferenciar atributo do parâmetro
    Termostato(){
        this->temperatura = 20; // inicia em 20 conforme o enunciado
        this->ligado = false;
    }

    // CORREÇÃO: destrutor não precisa de ponto e vírgula e pode ficar vazio
    ~Termostato(){}

    void ligar(){
        ligado = true;
    }

    void desligar(){
        ligado = false;
    }

    // CORREÇÃO: aumentar/diminuir recebem graus como parâmetro, conforme o enunciado
    void aumentar(int graus){
        if (!ligado){
            cout << "Aviso: Termostato desligado!" << endl;
            return;
        }

        int nova_temp = temperatura + graus;

        // CORREÇÃO: valida a nova temperatura ANTES de aplicar
        if (validar_Temperatura(nova_temp)){
            temperatura = nova_temp;
        }
    }

    void diminuir(int graus){
        if (!ligado){
            cout << "Aviso: Termostato desligado!" << endl;
            return;
        }

        int nova_temp = temperatura - graus;

        // CORREÇÃO: valida a nova temperatura ANTES de aplicar
        if (validar_Temperatura(nova_temp)){
            temperatura = nova_temp;
        }
    }

    void exibir_estado(){
        cout << "Ligado: " << (ligado ? "Sim" : "Nao") << endl;
        cout << "Temperatura: " << temperatura << "°C" << endl;
    }
};

int main()
{
    Termostato ar; // CORREÇÃO: construtor sem argumentos, não precisa passar nada

    // CORREÇÃO: faltavam os () em todas as chamadas de método
    ar.ligar();
    ar.exibir_estado();
    ar.aumentar(5);
    ar.aumentar(5);
    ar.exibir_estado();
    ar.diminuir(3);
    ar.exibir_estado();
    ar.desligar();
    ar.diminuir(3); // deve exibir aviso pois está desligado
    ar.exibir_estado();

    return 0;
}

File: python/Termostato.py
class Termostato:
    MIN = 0
    MAX = 40
 
    def __init__(self):
        self.__temperatura = 20  # inicia em 20 conforme o enunciado
        self.__ligado = False
 
    def __del__(self):
        pass  # destrutor vazio, equivalente ao ~Termostato(){}
 
    def __validar_temperatura(self, valor) -> bool:
        """Retorna True se a temperatura for válida, False se estiver fora dos limites."""
        if valor < self.MIN or valor > self.MAX:
            print("Aviso: Temperatura fora dos limites permitidos (0 a 40)!")
            return False
        return True
 
    def ligar(self):
        self.__ligado = True
 
    def desligar(self):
        self.__ligado = False
 
    def aumentar(self, graus: int):
        if not self.__ligado:
            print("Aviso: Termostato desligado!")
            return
 
        nova_temp = self.__temperatura + graus
 
        if self.__validar_temperatura(nova_temp):
            self.__temperatura = nova_temp
 
    def diminuir(self, graus: int):
        if not self.__ligado:
            print("Aviso: Termostato desligado!")
            return
 
        nova_temp = self.__temperatura - graus
 
        if self.__validar_temperatura(nova_temp):
            self.__temperatura = nova_temp
 
    def exibir_estado(self):
        print(f"Ligado: {'Sim' if self.__ligado else 'Nao'}")
        print(f"Temperatura: {self.__temperatura}°C")
 
 
if __name__ == "__main__":
    ar = Termostato()
 
    ar.ligar()
    ar.exibir_estado()
    ar.aumentar(5)
    ar.aumentar(5)
    ar.exibir_estado()
    ar.diminuir(3)
    ar.exibir_estado()
    ar.desligar()
    ar.diminuir(3)  
    ar.exibir_estado()